### ISSUE TITLE
[Phase 4] Quantile Regression for Uncertainty Estimation

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/demand-forecasting-engine#5]
## Summary

Extend models to produce prediction intervals via quantile regression, enabling uncertainty estimation for inventory planning. Instead of single-point forecasts, generate probabilistic predictions with lower and upper bounds.

## System Context

```
demand-forecasting-engine/
├── src/
│   ├── models/
│   │   ├── base_model.py           # EXISTS
│   │   ├── lightgbm_model.py       # EXISTS
│   │   └── quantile_model.py       # <- CREATE
│   └── evaluation/
│       ├── metrics.py              # EXISTS
│       └── interval_metrics.py     # <- CREATE
└── scripts/
    └── train_quantile.py           # <- CREATE
```

## Dependencies

- Requires: Phase 2 (Single Model Baseline), Phase 3 (Cross-Validation)
- Blocks: Phase 5 (Ensemble), Phase 10 (Production Pipeline)

### 1. Quantile LightGBM Model (`src/models/quantile_model.py`)

```python
import numpy as np
import pandas as pd
import lightgbm as lgb
from typing import List, Optional, Dict, Any, Union
from .base_model import BaseForecaster


class QuantileLightGBM(BaseForecaster):
    """
    LightGBM with quantile regression for uncertainty estimation.

    Trains separate models for each quantile to produce prediction intervals.

    Parameters
    ----------
    quantiles : list of float
        Quantiles to predict (e.g., [0.1, 0.5, 0.9] for 80% interval)
    n_estimators : int
        Number of boosting rounds per quantile model
    learning_rate : float
        Boosting learning rate
    max_depth : int
        Maximum tree depth
    num_leaves : int
        Maximum leaves per tree
    """

    def __init__(
        self,
        quantiles: List[float] = [0.1, 0.5, 0.9],
        n_estimators: int = 1000,
        learning_rate: float = 0.05,
        max_depth: int = -1,
        num_leaves: int = 31,
        min_child_samples: int = 20,
        early_stopping_rounds: int = 50,
        random_state: int = 42,
        n_jobs: int = -1,
        verbose: int = -1,
    ):
        self.quantiles = sorted(quantiles)
        self.n_estimators = n_estimators
        self.learning_rate = learning_rate
        self.max_depth = max_depth
        self.num_leaves = num_leaves
        self.min_child_samples = min_child_samples
        self.early_stopping_rounds = early_stopping_rounds
        self.random_state = random_state
        self.n_jobs = n_jobs
        self.verbose = verbose

        self.models_: Dict[float, lgb.Booster] = {}
        self.feature_names_: Optional[List[str]] = None

    def _get_lgb_params(self, quantile: float) -> Dict[str, Any]:
        """Get LightGBM parameters for a specific quantile."""
        return {
            'objective': 'quantile',
            'alpha': quantile,  # The quantile to predict
            'metric': 'quantile',
            'boosting_type': 'gbdt',
            'learning_rate': self.learning_rate,
            'max_depth': self.max_depth,
            'num_leaves': self.num_leaves,
            'min_child_samples': self.min_child_samples,
            'random_state': self.random_state,
            'n_jobs': self.n_jobs,
            'verbose': self.verbose,
        }

    def fit(
        self,
        X: pd.DataFrame,
        y: pd.Series,
        eval_set: Optional[tuple] = None,
    ) -> 'QuantileLightGBM':
        """
        Train quantile models.

        Trains one LightGBM model per quantile.
        """
        self.feature_names_ = list(X.columns)
        train_data = lgb.Dataset(X, label=y)

        valid_sets = [train_data]
        if eval_set is not None:
            X_val, y_val = eval_set
            valid_data = lgb.Dataset(X_val, label=y_val, reference=train_data)
            valid_sets.append(valid_data)

        callbacks = []
        if self.early_stopping_rounds and eval_set is not None:
            callbacks.append(lgb.early_stopping(self.early_stopping_rounds))

        for quantile in self.quantiles:
            if self.verbose >= 0:
                print(f"Training quantile {quantile}...")

            params = self._get_lgb_params(quantile)

            self.models_[quantile] = lgb.train(
                params,
                train_data,
                num_boost_round=self.n_estimators,
                valid_sets=valid_sets,
                callbacks=callbacks,
            )

        return self

    def predict(
        self,
        X: pd.DataFrame,
        return_all_quantiles: bool = True,
    ) -> Union[np.ndarray, Dict[float, np.ndarray]]:
        """
        Generate quantile predictions.

        Parameters
        ----------
        X : pd.DataFrame
            Feature matrix
        return_all_quantiles : bool
            If True, return 2D array with all quantiles.
            If False, return dict mapping quantile to predictions.

        Returns
        -------
        np.ndarray or dict
            If return_all_quantiles: shape (n_samples, n_quantiles)
            Otherwise: {quantile: predictions}
        """
        predictions = {}
        for quantile, model in self.models_.items():
            predictions[quantile] = model.predict(X)

        if return_all_quantiles:
            return np.column_stack([predictions[q] for q in self.quantiles])
        return predictions

    def predict_median(self, X: pd.DataFrame) -> np.ndarray:
        """Return median (0.5 quantile) prediction."""
        if 0.5 not in self.models_:
            raise ValueError("No median model. Add 0.5 to quantiles.")
        return self.models_[0.5].predict(X)

    def predict_interval(
        self,
        X: pd.DataFrame,
        lower_quantile: float = 0.1,
        upper_quantile: float = 0.9,
    ) -> tuple:
        """
        Return prediction interval.

        Returns
        -------
        tuple
            (lower_bound, upper_bound) arrays
        """
        if lower_quantile not in self.models_ or upper_quantile not in self.models_:
            raise ValueError(f"Quantiles {lower_quantile}, {upper_quantile} not trained.")

        lower = self.models_[lower_quantile].predict(X)
        upper = self.models_[upper_quantile].predict(X)
        return lower, upper

    def get_feature_importance(self, quantile: float = 0.5) -> pd.DataFrame:
        """Get feature importance for specified quantile model."""
        if quantile not in self.models_:
            quantile = self.quantiles[len(self.quantiles) // 2]

        importance = self.models_[quantile].feature_importance(importance_type='gain')
        return pd.DataFrame({
            'feature': self.feature_names_,
            'importance': importance
        }).sort_values('importance', ascending=False)

    def get_params(self) -> Dict[str, Any]:
        return {
            'quantiles': self.quantiles,
            'n_estimators': self.n_estimators,
            'learning_rate': self.learning_rate,
            'max_depth': self.max_depth,
            'num_leaves': self.num_leaves,
            'min_child_samples': self.min_child_samples,
        }

    def set_params(self, **params) -> 'QuantileLightGBM':
        for key, value in params.items():
            if hasattr(self, key):
                setattr(self, key, value)
        return self
```

### 2. Interval Metrics (`src/evaluation/interval_metrics.py`)

```python
import numpy as np
from typing import Tuple


def interval_coverage(
    y_true: np.ndarray,
    lower: np.ndarray,
    upper: np.ndarray,
) -> float:
    """
    Calculate prediction interval coverage.

    Coverage = fraction of true values within [lower, upper].

    Parameters
    ----------
    y_true : np.ndarray
        Actual values
    lower : np.ndarray
        Lower bound of interval
    upper : np.ndarray
        Upper bound of interval

    Returns
    -------
    float
        Coverage as decimal (0.80 = 80% coverage)
    """
    within_interval = (y_true >= lower) & (y_true <= upper)
    return within_interval.mean()


def interval_width(lower: np.ndarray, upper: np.ndarray) -> float:
    """Calculate mean interval width."""
    return (upper - lower).mean()


def interval_score(
    y_true: np.ndarray,
    lower: np.ndarray,
    upper: np.ndarray,
    alpha: float = 0.2,
) -> float:
    """
    Calculate interval score (Winkler score).

    Combines coverage and width into single metric.
    Lower is better.

    Parameters
    ----------
    y_true : np.ndarray
        Actual values
    lower, upper : np.ndarray
        Interval bounds
    alpha : float
        1 - desired coverage (e.g., 0.2 for 80% interval)

    Returns
    -------
    float
        Interval score
    """
    width = upper - lower

    # Penalty for under-prediction
    under = np.maximum(lower - y_true, 0)
    # Penalty for over-prediction
    over = np.maximum(y_true - upper, 0)

    score = width + (2 / alpha) * under + (2 / alpha) * over
    return score.mean()


def quantile_loss(
    y_true: np.ndarray,
    y_pred: np.ndarray,
    quantile: float,
) -> float:
    """
    Calculate pinball (quantile) loss.

    Parameters
    ----------
    y_true : np.ndarray
        Actual values
    y_pred : np.ndarray
        Predicted values
    quantile : float
        Quantile being predicted (0-1)

    Returns
    -------
    float
        Mean quantile loss
    """
    error = y_true - y_pred
    loss = np.maximum(quantile * error, (quantile - 1) * error)
    return loss.mean()


def check_quantile_monotonicity(predictions: np.ndarray) -> bool:
    """
    Check if quantile predictions are monotonically increasing.

    Parameters
    ----------
    predictions : np.ndarray
        Shape (n_samples, n_quantiles), columns in increasing quantile order

    Returns
    -------
    bool
        True if all q_i < q_{i+1} for all samples
    """
    for i in range(predictions.shape[1] - 1):
        if not np.all(predictions[:, i] <= predictions[:, i + 1]):
            return False
    return True


def fix_quantile_crossing(
    predictions: np.ndarray,
    method: str = 'sort'
) -> np.ndarray:
    """
    Fix quantile crossing by ensuring monotonicity.

    Parameters
    ----------
    predictions : np.ndarray
        Shape (n_samples, n_quantiles)
    method : str
        'sort': Sort predictions per row
        'average': Average crossing quantiles

    Returns
    -------
    np.ndarray
        Fixed predictions
    """
    if method == 'sort':
        return np.sort(predictions, axis=1)
    elif method == 'average':
        fixed = predictions.copy()
        for i in range(1, predictions.shape[1]):
            mask = fixed[:, i] < fixed[:, i-1]
            avg = (fixed[mask, i] + fixed[mask, i-1]) / 2
            fixed[mask, i] = avg
            fixed[mask, i-1] = avg
        return fixed
    else:
        raise ValueError(f"Unknown method: {method}")
```

### 3. Training Script (`scripts/train_quantile.py`)

```python
#!/usr/bin/env python
"""Train quantile regression model."""
import argparse
import numpy as np
import pandas as pd

from src.data.loader import load_sales_data
from src.features.pipeline import build_feature_pipeline
from src.models.quantile_model import QuantileLightGBM
from src.evaluation.splits import temporal_train_test_split
from src.evaluation.interval_metrics import (
    interval_coverage, interval_width, interval_score,
    check_quantile_monotonicity
)


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--data", required=True)
    parser.add_argument("--quantiles", default="0.1,0.5,0.9")
    args = parser.parse_args()

    quantiles = [float(q) for q in args.quantiles.split(",")]

    # Load and prepare
    df = load_sales_data(args.data)
    pipeline = build_feature_pipeline()
    df_features = pipeline.fit_transform(df)

    feature_cols = [c for c in df_features.columns
                    if c not in ['date', 'sku_id', 'sales']]
    X = df_features[feature_cols].dropna()
    y = df_features.loc[X.index, 'sales']

    X_train, X_test, y_train, y_test = temporal_train_test_split(X, y)

    # Train
    print(f"Training quantile model with q={quantiles}")
    model = QuantileLightGBM(quantiles=quantiles)
    model.fit(X_train, y_train, eval_set=(X_test, y_test))

    # Predict
    predictions = model.predict(X_test)
    q10, q50, q90 = predictions[:, 0], predictions[:, 1], predictions[:, 2]

    # Check monotonicity
    is_monotonic = check_quantile_monotonicity(predictions)
    print(f"\nQuantile monotonicity: {'PASS' if is_monotonic else 'FAIL'}")

    # Evaluate intervals
    coverage = interval_coverage(y_test.values, q10, q90)
    width = interval_width(q10, q90)
    score = interval_score(y_test.values, q10, q90, alpha=0.2)

    print(f"\n80% Prediction Interval Metrics:")
    print(f"  Coverage: {coverage:.1%} (target: 80%)")
    print(f"  Mean Width: {width:.2f}")
    print(f"  Interval Score: {score:.2f}")


if __name__ == "__main__":
    main()
```

## Implementation Checklist

- [ ] Implement `QuantileLightGBM` with multi-quantile training
- [ ] Add `predict_interval()` and `predict_median()` methods
- [ ] Implement interval coverage metric
- [ ] Implement interval score (Winkler score)
- [ ] Add quantile loss function
- [ ] Add monotonicity checking and fixing
- [ ] Create training script
- [ ] Visualize prediction intervals

## Files to Modify

| File | Lines | Action | Description |
|------|-------|--------|-------------|
| `src/models/quantile_model.py` | 0-180 | Create | Quantile regression model |
| `src/evaluation/interval_metrics.py` | 0-120 | Create | Interval evaluation metrics |
| `scripts/train_quantile.py` | 0-60 | Create | Training script |

## Technical Challenges

**Understanding Quantile Loss (Pinball Loss)**:

```python
# For quantile q:
# - Under-prediction penalized by q
# - Over-prediction penalized by (1-q)

# q=0.1: Mostly penalizes over-prediction (want low estimates)
# q=0.9: Mostly penalizes under-prediction (want high estimates)

def pinball_loss(y_true, y_pred, q):
    error = y_true - y_pred
    return max(q * error, (q-1) * error)

# Example: y_true=100, y_pred=80, q=0.9
# error = 100 - 80 = 20 (under-predicted)
# loss = max(0.9 * 20, -0.1 * 20) = max(18, -2) = 18
```

**Quantile Crossing**: When q10 > q50 or q50 > q90. Fix by sorting or training jointly.

## Definition of Done

- [ ] Model produces predictions for all specified quantiles
- [ ] No quantile crossing (monotonicity check passes)
- [ ] 80% interval achieves ~80% coverage (75-85%)
- [ ] Interval score calculated and logged
- [ ] Tests pass: `pytest tests/test_quantile.py -v`

## Risk Assessment

| Risk | Impact | Mitigation |
|------|--------|------------|
| Quantile crossing | :yellow_circle: Medium | Post-process with sorting |
| Poor calibration | :yellow_circle: Medium | Tune quantile model separately |
| Slow training (3x models) | :green_circle: Low | Parallel training or fewer quantiles |

## Time Estimate

**Estimated Hours**: 8-10 hours
**Priority**: High (required for uncertainty estimation)
